File: Course-1/Cpp-Projects/Data-Structures-and-Algorithms-2024/Assignment-9/task1.cpp
/*
Given an oriented graph, determine if it has a cycle of negative weight and print it (if exists).

Input
Input's first line has number N (1 ≤ N ≤ 100) — number of vertices in the graph. Each of the next N lines contains N numbers — all representing an adjacency matrix. All weights are strictly less than 10000 by absolute value. If there is no edge, corresponding value will be exactly 100000.

Output
First line should YES if there exists a negative cycle, or NO otherwise. If the cycle exists, second line should should contain the number of vertices in that cycle and third line should contain indices of those vertices (in cycle order).

Input
2
0 -1
-1 0

Output
YES
2
2 1 
*/
#include <string>
#include <stack>
#include <queue>
#include <algorithm>
#include <set>
#include <map>
#include <list>
#include <cmath>
#include <vector>
#include <memory>
#include <iostream>
#include <iomanip>
#include <stdio.h>
#include <cmath>
using namespace std;
using ll = long long;
using ull = unsigned long long;
using db = double;
using uint = unsigned int;


const int INF = 1e9;


template <typename T> class Vertex {
public:
    T key;
    int index;
    Vertex(T key, int index) {
        this->key = key;
        this->index = index;
    }
};


template <typename V, typename  E> class Edge {
public:
    E weight;
    shared_ptr<Vertex<V>> from;
    shared_ptr<Vertex<V>> to;
    Edge(shared_ptr<Vertex<V>> from, shared_ptr<Vertex<V>> to, E weight) {
        this->from = from;
        this->to = to;
        this->weight = weight;
    }
};


template <typename V, typename E> class Graph {
public:
    list<shared_ptr<Vertex<V>>> vertices;
    list<shared_ptr<Edge<V, E>>> edges;
    int size;
    shared_ptr<Vertex<V>> getVertex(V key) {
        for (auto it = this->vertices.begin(); it != this->vertices.end(); it++) {
            shared_ptr<Vertex<V>> vertex = *it;
            if (vertex->key == key) {
                return vertex;
            }
        }
        return nullptr;
    }
    void addVertex(shared_ptr<Vertex<V>> vertex) {
        this->vertices.push_back(vertex);
        this->size++;
    }
    void addEdge(shared_ptr<Edge<V, E>> edge) {
        this->edges.push_back(edge);
    }
    void newVertex(V key) {
        shared_ptr<Vertex<V>> vertex = make_shared<Vertex<V>>(key, this->size++);
        this->vertices.push_back(vertex);
    }
    void newEdge(V from, V to, E weight) {
        shared_ptr<Vertex<V>> fromVertex = getVertex(from);
        shared_ptr<Vertex<V>> toVertex = getVertex(to);
        shared_ptr<Edge<V, E>> edge = make_shared<Edge<V, E>>(fromVertex, toVertex, weight);
        this->edges.push_back(edge);
    }
    int getSize() {
        return this->size;
    }
    bool relax(shared_ptr<Edge<V, E>> edge, shared_ptr<int[]> distance, shared_ptr<int[]> path) {
        if (distance[edge->to->index] > distance[edge->from->index] + edge->weight) {
            distance[edge->to->index] = distance[edge->from->index] + edge->weight;
            path[edge->to->index] = edge->from->index;
            return true;
        }
        return false;
    }
    void printSycle(shared_ptr<int[]> path, int last) {
        stack<int> cycle;
        for (int v = path[last]; v != last; v = path[v]) {
            cycle.push(v);
        }
        cycle.push(last);
        cout << cycle.size() << "\n";
        while (!cycle.empty()) {
            int vertex = cycle.top();
            cout << vertex + 1 << " ";
            cycle.pop();
        }
        cout << "\n";
    }
    bool AndreyTorgashinov_sp(int s) {
        shared_ptr<int[]> distance(new int[this->size]);
        shared_ptr<int[]> path(new int[this->size]);
        for (int i = 0; i < this->size; i++) {
            if (i != s) {
                distance[i] = INF;
            }
            else {
                distance[i] = 0;
            }
            path[i] = i;
        }
        for (int i = 0; i < this->size; i++) {
            for (auto it = this->edges.begin(); it != this->edges.end(); it++) {
                shared_ptr<Edge<V, E>> edge = *it;
                if (distance[edge->from->index] != INF) {
                    relax(edge, distance, path);
                }
            }
        }
        for (auto it = this->edges.begin(); it != this->edges.end(); it++) {
            shared_ptr<Edge<V, E>> edge = *it;
            if (distance[edge->from->index] != INF) {
                if (relax(edge, distance, path)) {
                    cout << "YES\n";
                    int from = edge->from->index;
                    for (int i = 0; i < this->size - 1; i++) {
                        from = path[from];
                    }
                    printSycle(path, from);
                    return false;
                }
            }
        }
        return true;
    }
    Graph() {
        this->size = 0;
    }
};


int main() {
    ios::sync_with_stdio(0);
    cin.tie();
    Graph<int, int> graph;
    int n;
    cin >> n;
    for (int i = 1; i <= n; i++) {
        graph.newVertex(i);
    }
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= n; j++) {
            int weight;
            cin >> weight;
            if (weight != 100000) {
                graph.newEdge(i, j, weight);
            }
        }
    }
    int i;
    for (i = 0; i < n; i++) {
        if (!graph.AndreyTorgashinov_sp(i)) {
            break;
        }
    }
    if (i == n) {
        cout << "NO\n";
    }
    return 0;
}

File: Course-1/Java-Projects/Data-Structures-and-Algorithms-2024/Assignment-8/task1.java
/*
A road map for a number of cities is given as an adjacency matrix. Write a program that will be able to tell if any city is reachable from any other city (perhaps via other cities).

Входные данные
First line has the number of cities on a map N (1 ≤ N ≤ 1000). Each of the next N lines has N numbers (0 or 1, separated with spaces) – elements of the adjacency matrix for the graph corresponding to the road map.

Выходные данные
Your program should output YES if every city is reachable from every other city, or NO otherwise.

Входные данные
5
0 0 1 0 0
0 0 1 0 1
1 1 0 0 0
0 0 0 0 0
0 1 0 0 0

Выходные данные
NO
*/
import java.util.LinkedList;
import java.util.Scanner;


public class Main {
    public static void main(String[] args) {
        int n;
        Scanner s = new Scanner(System.in);
        Graph<Integer, Integer> graph = new Graph<>();
        n = s.nextInt();
        for (int i = 0; i < n; i++) {
            graph.newVertex(i);
        }
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (s.nextInt() > 0) {
                    graph.newEdge(0, graph.getVertex(i), graph.getVertex(j));
                }
            }
        }
        int[] components = new int[n];
        int numberOfComponents = 1;
        for (int i = 0; i < n; i++) {
            if (components[i] == 0) {
                graph.AndreyTorgashinov_dfs(graph.getVertex(i), components, numberOfComponents);
                numberOfComponents++;
            }
        }
        if (numberOfComponents == 2) {
            System.out.println("YES");
        } else{
            System.out.println("NO");
        }
    }
}


class Edge <T> {
    protected T weight;
    protected Vertex from;
    protected Vertex to;
    Edge(Graph graph, Vertex from, Vertex to, T weight) {
        this.from = from;
        this.to = to;
        this.weight = weight;
        from.edges.add(this);
    }
}


class Vertex <T> {
    protected LinkedList<Edge> edges;
    protected T key;
    protected int index;
    Vertex(Graph graph, T key) {
        edges = new LinkedList<>();
        this.key = key;
        this.index = graph.getSize();
    }
}


class Graph <V, E> {
    private LinkedList<Vertex> vertices;
    private LinkedList<Edge> edges;
    private int size;
    public void newVertex(V key) {
        this.vertices.add(new Vertex(this, key));
        this.size++;
    }
    public void newEdge(E weight, Vertex from, Vertex to) {
        this.edges.add(new Edge(this, from, to, weight));
    }
    public Vertex getVertex(V key) {
        for (Vertex v : this.vertices) {
            if (v.key == key) {
                return v;
            }
        }
        return null;
    }
    public int getSize() {
        return this.size;
    }
    public void AndreyTorgashinov_dfs(Vertex v, int[] components, int numberOfComponents) {
        components[v.index] = numberOfComponents;
        for (Object edge : v.edges) {
            Vertex u = ((Edge) edge).to;
            if (components[u.index] == 0) {
                AndreyTorgashinov_dfs(u, components, numberOfComponents);
            }
        }
    }
    Graph() {
        this.vertices = new LinkedList<>();
        this.edges = new LinkedList<>();
        this.size = 0;
    }
}